packages/react/src/elements/components/Icon/Icon.js

```diff
@@ -1,76 +1,124 @@
+/** @type {string[]} */
+const availableSizes = Object.values(sizes);
 
-const AvailableSizes = ["24", "16"];
+/**
+ * @typedef {Object} IconProps
+ * @property {string} name
+ * @property {string} svg
+ * @property {string} size
+ * @property {string} [nameOrSVG]
+ */
 
-export default class Icon extends Component {
-  setSizedIcon(icon, size) {
-    if (!(icon && size)) {
-      return; // Silently return until name and size are set
-    }
+/**
+ * @param {string} size
+ * @returns {boolean}
+ */
+function isValidSize(size) {
+  return availableSizes.includes(size);
+}
 
-    return AvailableSizes.indexOf(size) > -1
-      ? this._confirmNameOrSVG(icon, size)
-      : console.error(
-          `Icon named "${icon} size "${size}" not found, only these size are allowed: `,
-          AvailableSizes
-        );
-  }
+/**
+ * @param {string} name
+ * @param {string} size
+ * @returns {string} SVG markup
+ */
+function getIconMarkupFromName(name, size) {
+  const key = `${name}-${size}`;
+
+  return Icons[key];
+}
+
+/**
+ * @param {string} nameOrSVG
+ * @param {string} size
+ * @returns {string} SVG markup
+ * @todo Disable logging in production
+ */
+function getIconMarkupFromOverload(nameOrSVG, size) {
+  const isSVG = /^<svg/.test(nameOrSVG);
+
+  return isSVG ? nameOrSVG : getIconMarkupFromName(nameOrSVG, size);
+}
+
+/**
+ * @param {IconProps} props
+ * @returns {string|undefined} SVG markup, or `undefined` when no icon is available
+ * @todo Disable logging in production
+ */
+function getIconMarkup({ name, svg, nameOrSVG, size }) {
+  if (svg) return svg;
+  if (!isValidSize(size)) return undefined;
+  if (name) return getIconMarkupFromName(name, size);
+  if (nameOrSVG) return getIconMarkupFromOverload(nameOrSVG, size);
+
+  return undefined;
+}
 
-  setIconClass = (icon, size) =>
-    AvailableSizes.includes(size)
-      ? `hig__icon--${size}-size`
-      : console.error(
-          `Icon named "${icon} size "${size}" not found, only these size are allowed: `,
-          AvailableSizes
-        );
-
-  _confirmNameOrSVG = (icon, size) => {
-    const isNamedIcon = Icons[`${icon}-${size}`];
-    const isSVG = /^<svg/.test(icon);
-
-    if (isNamedIcon) {
-      return isNamedIcon;
-    } else if (isSVG) {
-      return icon;
-    }
-    console.warn(`NO HIG ICON: ${icon}`);
-  };
-
-  render() {
-    const iconClasses = cx(
-      "hig__icon",
-      this.setIconClass(this.props.nameOrSvg, this.props.size)
-    );
-
-    const iconString = this.setSizedIcon(this.props.nameOrSVG, this.props.size);
-    return (
-      <div
-        className={iconClasses}
-        dangerouslySetInnerHTML={{ __html: iconString }}
-      />
-    );
+/**
+ * @returns {JSX.Element}
+ */
+function IconPlaceholder() {
+  return <div className={COMPONENT_CLASS} />;
+}
+
+/**
+ * @param {IconProps} props
+ * @returns {JSX.Element}
+ * @todo Disable logging in production
+ */
+export default function Icon(props) {
```

Comment:
I like having the logic flow more clearly. this makes a lot of sense as far as validating whether an icon exists.